ron/lib/RON/Schema/EDN.hs

```diff
@@ -0,0 +1,228 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module RON.Schema.EDN (parseSchema) where
+
+import           RON.Internal.Prelude
+
+import           Control.Monad.State.Strict (StateT, evalStateT, get, gets, put)
+import           Control.Monad.Trans (MonadTrans, lift)
+import           Control.Monad.Trans.Identity (runIdentityT)
+import           Data.Attoparsec.Lazy (Result (Done))
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy.Char8 as BSLC
+import           Data.Char (isSpace)
+import           Data.EDN (Tagged (NoTag, Tagged), Value (List, Map, Symbol),
+                           (.!=), (.:?))
+import           Data.EDN.Encode (fromTagged)
+import           Data.EDN.Parser (parseBSL)
+import           Data.EDN.Types (EDNList, EDNMap)
+import           Data.EDN.Types.Class (Parser, parseEither, typeMismatch)
+import qualified Data.Map.Strict as Map
+import qualified Data.Text as Text
+import qualified Data.Text.Encoding as Text
+import qualified Data.Text.Lazy as TextL
+import           Data.Text.Lazy.Builder (toLazyText)
+import qualified Data.Text.Lazy.Encoding as TextL
+
+import           RON.Schema
+
+newtype Env = Env {knownTypes :: Map Text RonType}
+    deriving (Show)
+
+startEnv :: Env
+startEnv = Env
+    { knownTypes = Map.fromList
+        [ ("AtomInteger",   atomInteger)
+        , ("AtomString",    atomString)
+        , ("RgaString",     rgaString)
+        , ("VersionVector", versionVector)
+        ]
+    }
+
+type Parser' = StateT Env Parser
+
+parseDeclaration :: Tagged Value -> Parser' Declaration
+parseDeclaration = withNoTag $ withList "declaration" $ \case
+    func : args -> do
+        func' <- withNoTag pure func
+        withSymbol "declaration name symbol" (go args) func'
+    [] -> fail "empty declaration"
+  where
+    go args = withNoPrefix "declaration name" $ \case
+        "struct_lww" -> DStructLww <$> parseStructLww args
+        name         -> fail $ "unknown declaration " ++ decodeUtf8 name
+
+parseStructLww :: EDNList -> Parser' StructLww
+parseStructLww code = do
+    struct@StructLww{structName} <- case code of
+        name : body -> do
+            let (annotations, fields) = span isTagged body
+            StructLww
+                <$> parseName name
+                <*> parseFields fields
+                <*> parseAnnotations annotations
+        _ -> fail
+            "Expected declaration in the form\
+            \ (struct_lww <name:symbol> <annotations>... <fields>...)"
+    env@Env{knownTypes} <- get
+    case Map.lookup structName knownTypes of
+        Nothing ->
+            put env { knownTypes =
+                        Map.insert structName (structLww struct) knownTypes
+                    }
+        Just _ ->
+            fail $ "duplicate declaration of type " ++ Text.unpack structName
+    pure struct
+
+  where
+
```

Comment:
Did empty line rest for a comment?

Reply:
No, they are for the visual balance with empty lines between functions inside `where` block.

Reply:
I offer to leave an empty line  before `where`, not after one. Here and feather.

Reply:
I don't like it. Maybe vice versa? Anyway, bikeshedding.